src/soc/ibm/power9/istep_8_6.c

```diff
@@ -0,0 +1,280 @@
+	uint32_t i;
+
+	/* Processor SMP Fabric broadcast scope configuration.
+	 *    CHIP_IS_NODE = MODE1 = default
+	 *    CHIP_IS_GROUP = MODE2
+	 */
+	uint8_t pump_mode = CHIP_IS_GROUP; /* ATTR_PROC_FABRIC_PUMP_MODE MODE2 from talos.xml */
+
+	switch(eps_table_type)
+	{
+	case EPS_TYPE_HE:
+		if (pump_mode == CHIP_IS_NODE) {
+			eps_r[0] = EPSILON_R_T0_HE[*floor_ratio];
+			eps_r[1] = EPSILON_R_T1_HE[*floor_ratio];
+			eps_r[2] = EPSILON_R_T2_HE[*floor_ratio];
+
+			eps_w[0] = EPSILON_W_T0_HE[*floor_ratio];
+			eps_w[1] = EPSILON_W_T1_HE[*floor_ratio];
+		} else {
+			eps_r[0] = EPSILON_R_T0_F4[*floor_ratio];
+			eps_r[1] = EPSILON_R_T1_F4[*floor_ratio];
+			eps_r[2] = EPSILON_R_T2_F4[*floor_ratio];
+
+			eps_w[0] = EPSILON_W_T0_F4[*floor_ratio];
+			eps_w[1] = EPSILON_W_T1_F4[*floor_ratio];
+		}
+
+		break;
+	case EPS_TYPE_HE_F8:
+		eps_r[0] = EPSILON_R_T0_F8[*floor_ratio];
+
+		if (pump_mode == CHIP_IS_NODE)
+			eps_r[1] = EPSILON_R_T1_F8[*floor_ratio];
+		else
+			eps_r[1] = EPSILON_R_T0_F8[*floor_ratio];
+
+		eps_r[2] = EPSILON_R_T2_F8[*floor_ratio];
+
+		eps_w[0] = EPSILON_W_T0_F8[*floor_ratio];
+		eps_w[1] = EPSILON_W_T1_F8[*floor_ratio];
+		break;
+	case EPS_TYPE_LE:
+		eps_r[0] = EPSILON_R_T0_LE[*floor_ratio];
+
+		if (pump_mode == CHIP_IS_NODE)
+			eps_r[1] = EPSILON_R_T1_LE[*floor_ratio];
+		else
+			eps_r[1] = EPSILON_R_T0_LE[*floor_ratio];
+
+		eps_r[2] = EPSILON_R_T2_LE[*floor_ratio];
+
+		eps_w[0] = EPSILON_W_T0_LE[*floor_ratio];
+		eps_w[1] = EPSILON_W_T1_LE[*floor_ratio];
+		break;
+	default:
+		printk(BIOS_WARNING, "Invalid epsilon table type 0x%.8X", eps_table_type);
+		break;
+	}
+
+	/* dump base epsilon values */
+	printk(BIOS_DEBUG, "Base epsilon values read from table:\n");
+	dump_eplisons();
+
+	eps_guardband = 20;
+
+	/* scale base epsilon values if core is running 2x nest frequency */
+	if (*ceiling_ratio == FABRIC_CORE_CEILING_RATIO_RATIO_8_8)
+	{
+		printk(BIOS_DEBUG, "Scaling based on ceiling frequency");
+		uint8_t scale_percentage = 100 * (*freq_ceiling) / (2 * (*fbc_freq));
+		scale_percentage -= 100;
+
+		/* scale/apply guardband read epsilons */
+		for (i = 0; i < NUM_EPSILON_READ_TIERS; i++)
+			config_guardband_epsilon(scale_percentage, &eps_r[i]);
+
+		/* Scale write epsilons */
+		for (i = 0; i < NUM_EPSILON_WRITE_TIERS; i++)
+			config_guardband_epsilon(scale_percentage, &eps_w[i]);
+	}
+
+	printk(BIOS_DEBUG, "Scaled epsilon values based on %s%d percent guardband:\n",
+			   (eps_guardband >= 0) ? ("+") : ("-"), eps_guardband);
+	dump_eplisons();
+
+	/* 
+	 * check relationship of epsilon counters:
+	 *   read tier values are strictly increasing
+	 *   write tier values are strictly increaing
+	 */
+	if ((eps_r[0] > eps_r[1]) || (eps_r[1] > eps_r[2]) || (eps_w[0] > eps_w[1]))
+		printk(BIOS_WARNING, "Invalid relationship between base epsilon values\n");
+}
+
+void configure_powerbus_link (void)
+{
+
+	if (!p9_calculate_frequencies(&core_floor_ratio, &core_ceiling_ratio,
+					&freq_fbc, &freq_core_ceiling)) {
```

Comment:
freq_fbc is uninitialized here, it is later compared in p9_calculate_frequencies() and never written to.